NFC-CTAP1 integration [DRAFT] [DO NOT MERGE]

krkhan opened the pull request:
Draft PR, contains lot of cruft that needs to be removed before actual review

ia0 committed:
Make store operations constant wrt flash operations

ia0 committed:
Test with nightly

ia0 committed:
Add unit-test for Store::entries

krkhan committed:
Add support for NFC transport

ia0 committed:
Return error instead of debug assert

With dirty storage we hit the assert. Returning an error permits to continue to
catch if the invariant is broken for normal operation while being able to
continue fuzzing with dirty storage.

kaczmarczyck committed:
updates status codes for RD02

kaczmarczyck committed:
renames residential to resident

kaczmarczyck committed:
Merge branch 'master' into new-status-codes

ia0 committed:
Add erase_storage application example

kaczmarczyck committed:
adds the new command AuthenticatorLargeBlobs

kaczmarczyck committed:
Merge pull request #245 from kaczmarczyck/new-status-codes

Updates status codes for RD02 of CTAP 2.1

ia0 committed:
Merge branch 'develop' into erase_storage

kaczmarczyck committed:
removes all occurences of CTAP 2.1 flags

kaczmarczyck committed:
removes all occurences of CTAP 2.1 flags from workflows

ia0 committed:
Merge pull request #247 from ia0/erase_storage

Add erase_storage application example

kaczmarczyck committed:
adds the config command

kaczmarczyck committed:
adds comments to new config command file

kaczmarczyck committed:
Merge pull request #250 from kaczmarczyck/workflows-2-1

removes all occurences of CTAP 2.1 flags from workflows

kaczmarczyck committed:
small readbility fix for variable assignment with cfg

kaczmarczyck committed:
Merge branch 'develop' into remove-flag-2-1

kaczmarczyck committed:
Merge pull request #252 from kaczmarczyck/remove-flag-2-1

Remove CTAP 2.1 flag

kaczmarczyck committed:
moves COSE related conversion from crypto to data_formats

kaczmarczyck committed:
adds storage changes for credential management

kaczmarczyck committed:
only keeps keys instead of credentials as state

kaczmarczyck commented on third_party/libtock-drivers/src/ctap_transport.rs:
```diff
@@ -0,0 +1,108 @@
+    fn setup(&self) -> bool;
+    fn recv(&self, buf: &mut [u8; 64]) -> bool;
+    fn send(&self, buf: &mut [u8; 64]) -> bool;
+    fn send_or_recv(&self, buf: &mut [u8; 64]) -> SendOrRecvStatus;
```
Do we actually use all these functions?

kaczmarczyck commented on third_party/libtock-drivers/src/ctap_transport.rs:
```diff
@@ -0,0 +1,108 @@
+static NFC_TRANSPORT: NfcTag = NfcTag {};
+static USB_TRANSPORT: UsbTransport = UsbTransport {};
+static mut TRANSPORT_STATE: TransportState = TransportState::Uninitialized;
+static mut CURRENT_TRANSPORT: &'static dyn Transport = &USB_TRANSPORT;
```
Why is this feature implemented with globals?

kaczmarczyck commented on third_party/libtock-drivers/src/nfc.rs:
```diff
@@ -24,6 +27,239 @@ mod allow_nr {
+    }
+
+    /// Function to identify the time elapsed for a transmission request.
+    fn bench_transmit(
```
Why is the benchmark in libtock now?

kaczmarczyck reviewed (commented):
Some general questions for this PR.

Also: Don't forget to request against the branch `develop` for the real PR!

kaczmarczyck committed:
moves filter_credentials to call side

kaczmarczyck committed:
Merge pull request #254 from kaczmarczyck/storage-management

Storage management

kaczmarczyck committed:
Merge branch 'develop' into cose-api

kaczmarczyck committed:
new CoseKey data format

kaczmarczyck committed:
Merge branch 'develop' into authenticator-config

kaczmarczyck committed:
updates license header to 2021 in ctap

kaczmarczyck committed:
code cleanups and clarifications

kaczmarczyck committed:
comment for constants

kaczmarczyck committed:
removes unused import

kaczmarczyck committed:
adds the command logic for credential management

kaczmarczyck committed:
moves constants to CoseKey

kaczmarczyck committed:
full URL

kaczmarczyck committed:
Merge pull request #253 from kaczmarczyck/cose-api

COSE conversion for ECDH and ECDSA

jmichelp committed:
Add Feitian OpenSK USB Dongle (#257) (#259)

Co-authored-by: superskybird <skybird.le@gmail.com>

Co-authored-by: Geoffrey <geoffrey@ftsafe.com>
Co-authored-by: superskybird <skybird.le@gmail.com>

kaczmarczyck committed:
Merge branch 'develop' into authenticator-config

kaczmarczyck committed:
Merge pull request #246 from kaczmarczyck/authenticator-config

Authenticator config: SetMinPinLength

kaczmarczyck committed:
Merge develop into command-cred-mgmt

kaczmarczyck committed:
removes unused duplicate PIN protocol check helper

kaczmarczyck committed:
renames to resident in README

kaczmarczyck committed:
Merge pull request #260 from kaczmarczyck/resident-naming

Resident naming

kaczmarczyck committed:
Merge branch 'develop' into command-cred-mgmt

kaczmarczyck committed:
implements Default for Response type

ia0 committed:
Merge branch 'develop' into v2_optim

ia0 committed:
Add bound-test in addition to equality-test

ia0 committed:
Merge pull request #238 from ia0/v2_optim

Make store operations constant wrt flash operations

ia0 committed:
Merge pull request #262 from ia0/v2_error

Return error instead of debug assert

kaczmarczyck committed:
only stores the RP ID index as state

kaczmarczyck committed:
reworks command state to its own struct

kaczmarczyck committed:
Merge branch 'develop' into command-cred-mgmt

kaczmarczyck committed:
stricter API for StatefulCommandPermission

kaczmarczyck committed:
fix typo

kaczmarczyck committed:
Merge pull request #256 from kaczmarczyck/command-cred-mgmt

Command logic for credential management

ia0 committed:
Make StoreUpdate generic over the byte slice ownership

This permits to call it without having to create a Vec<u8> when possible.

ia0 committed:
Merge pull request #264 from ia0/store_update_ref

Make StoreUpdate generic over the byte slice ownership

kaczmarczyck committed:
implements the credBlob extensions

kaczmarczyck committed:
Merge pull request #265 from kaczmarczyck/cred-blob-extension

credBlob extension

kaczmarczyck committed:
force PIN changes

kaczmarczyck committed:
makes test_get_info more readable

kaczmarczyck committed:
remove resolved TODO

kaczmarczyck committed:
use transactions, and how to store a bool

kaczmarczyck committed:
Merge pull request #267 from kaczmarczyck/force-pin-change

Force PIN change

kaczmarczyck committed:
bigger shards, fixed get_large_blob

kaczmarczyck committed:
large blob commit is one transaction

kaczmarczyck committed:
fixes const usage in test_get_info

kaczmarczyck committed:
improvements to large blob storage

ia0 committed:
Add fragmentation support

ia0 committed:
Remove useless check

ia0 committed:
Merge pull request #270 from ia0/fragmentation

Add helper functions for fragmented entries

kaczmarczyck committed:
Merge branch 'develop' into large-blobs

ia0 committed:
Improve count_credentials by not deserializing them

ia0 committed:
Factor test tools between store and fragment

Those need the driver to deal with the fact that the store is stateful. Those
tests can't be moved to the test suite because they use private functions.

kaczmarczyck committed:
use new store fragments

kaczmarczyck committed:
formats code, clippy

ia0 committed:
Merge pull request #271 from ia0/count_credentials

Improve count_credentials by not deserializing them

ia0 committed:
Merge branch 'develop' into factor_test

kaczmarczyck committed:
Merge branch 'develop' into large-blobs

kaczmarczyck committed:
style fix

kaczmarczyck committed:
reduce testing to not account for shard size

ia0 committed:
Merge pull request #272 from ia0/factor_test

Factor test tools between store and fragment

kaczmarczyck committed:
Merge branch 'develop' into large-blobs

kaczmarczyck committed:
adds enterprise for make, byte keys

kaczmarczyck committed:
new extension entry for largeBlobKey

kaczmarczyck committed:
Merge pull request #263 from kaczmarczyck/large-blobs

Large blobs

kaczmarczyck committed:
Merge pull request #273 from kaczmarczyck/commands-2-1

Updates command inputs to 2.1, unified u8 usage for map keys

kaczmarczyck committed:
Merge pull request #274 from kaczmarczyck/extension-large-blobs

largeBlobKey extension

kaczmarczyck committed:
remove conditional trait implementation

kaczmarczyck committed:
remove derive_debug feature

kaczmarczyck committed:
fix crypto workflow

kaczmarczyck committed:
Merge pull request #278 from kaczmarczyck/no-cfg-derive

Remove derive_debug feature

kaczmarczyck committed:
changes the handling of 0 credentials

kaczmarczyck committed:
allow RP ID permissions for some subcommands

kaczmarczyck committed:
dummy implementation for enterprise attestation

kaczmarczyck committed:
adds some documenation for enterprise attestation

kaczmarczyck committed:
correct const arrays of strings

kaczmarczyck committed:
Merge pull request #279 from kaczmarczyck/no-credential-rp

No credentials in enumerateRp

kaczmarczyck committed:
Merge branch 'develop' into resolve-permission

kaczmarczyck committed:
fix string conversion style

kaczmarczyck committed:
Merge branch 'develop' into enterprise-attestation

kaczmarczyck committed:
rename require_ to ensure

kaczmarczyck committed:
implements alwaysUv and makeCredUvNotRqd

kaczmarczyck committed:
adds changes to README

kaczmarczyck committed:
conditional allow instead of cfg not

kaczmarczyck committed:
Merge pull request #280 from kaczmarczyck/resolve-permission

Allow RP ID in token permissions for some subcommands of credential management

kaczmarczyck committed:
Merge branch 'develop' into enterprise-attestation

kaczmarczyck committed:
new test for attestation configuration

kaczmarczyck committed:
assert correct const usage in code

kaczmarczyck committed:
Merge pull request #281 from kaczmarczyck/enterprise-attestation

Enterprise attestation

kaczmarczyck committed:
changes always uv constant to a clearer version

kaczmarczyck committed:
update README to new constant

kaczmarczyck committed:
Merge branch 'develop' into always-uv

kaczmarczyck committed:
move some logic into storage.rs

kaczmarczyck committed:
Merge pull request #283 from kaczmarczyck/always-uv

Add the option alwaysUv

kaczmarczyck committed:
Improved documentation for customization (#287)

* move constants to new file, and update documentation

* documentation improvements, deploy checks tests

* fix pylint

* improved code style

* swap build and check

kaczmarczyck committed:
HKDF for CTAP2.1 (#290)

* implements hkdf, both regular and FIDO specific

* improved documentation

* constant usage in function return type

kaczmarczyck committed:
renames PIN protocol related variables to prepare PIN protocol v2 (#291)

kaczmarczyck committed:
adds the PIN protocol trait (#292)

* adds the PIN protocol trait

* improved documentation

* SharedSecret not mutable

ia0 committed:
Update the documentation to use linking by name

See https://doc.rust-lang.org/stable/rustdoc/linking-to-items-by-name.html

ia0 committed:
Fix typography

kaczmarczyck committed:
PIN protocol V2 in ClientPin (#293)

* PIN protocol V2 in ClientPin

* the test ClientPin has a random second private key

kaczmarczyck committed:
Merge branch 'develop' into doc

ia0 committed:
Merge pull request #294 from ia0/doc

Update the documentation to use linking by name

kaczmarczyck committed:
adds PIN protocol V2 to all commands (#295)

kaczmarczyck committed:
adds PIN token state with timeouts (#296)

kaczmarczyck committed:
Output parameters for CTAP2.1 (#297)

* finalizes output parameters for CTAP2.1

* explanation for internal UV
